### PR TITLE
Rework and simplify BriefcaseManager and BriefcaseDb

### PR DESCRIPTION
**Pull request by kabentley**
Note this PR contains breaking changes to the @beta api BriefcaseManager.download and @internal api NativeApp

### CODE REVIEW

File: editor/common/src/EditorProps.ts
Comment (pmconne) [inline editor/common/src/EditorProps.ts:39]: Missing docs and release tags.

File: full-stack-tests/core/src/frontend/standalone/EditTool.test.ts
Comment (pmconne) [inline full-stack-tests/core/src/frontend/standalone/EditTool.test.ts:50]: Remove `.only`.

File: test-apps/display-test-app/src/frontend/Viewer.ts
Comment (pmconne) [inline test-apps/display-test-app/src/frontend/Viewer.ts:409]: Did I misread or is `key` a hashed filename and so this will no longer work?

**Review by pmconne — CHANGES_REQUESTED** [review]

File: core/backend/src/IModelDb.ts
Comment (kabentley) [reply core/backend/src/IModelDb.ts:2428]: I moved that event to IModelDb

File: core/backend/src/rpc-impl/IModelWriteRpcImpl.ts
Comment (kabentley) [reply core/backend/src/rpc-impl/IModelWriteRpcImpl.ts]: done. Maybe there's some template way to do this?

File: core/backend/src/rpc-impl/NativeAppRpcImpl.ts
Comment (kabentley) [reply core/backend/src/rpc-impl/NativeAppRpcImpl.ts:48]: the documentation should be on the Rpc definition and not copied/repeated on the implementation. Everything I removed was redundant copies, not that much of it was useful anyway,

File: test-apps/display-test-app/src/frontend/Viewer.ts
Comment (kabentley) [reply test-apps/display-test-app/src/frontend/Viewer.ts]: You're right, I removed this because openStandaloneIModel should now work with the "same file" (it returns the existing Db). But, i don't know how to test this.

File: core/backend/src/IModelDb.ts
Comment (kabentley) [reply core/backend/src/IModelDb.ts]: it was, but no longer is.

File: test-apps/display-test-app/src/frontend/Viewer.ts
Comment (pmconne) [reply test-apps/display-test-app/src/frontend/Viewer.ts]: I added this because previously the following would produce an error - something from Raman's briefcase cache IIRC, so possibly fixed by your changes:
1. Open an iModel.
2. Click "open iModel" toolbar button and choose same iModel.
3. Observe cryptic error; iModel fails to open.

File: core/backend/src/rpc-impl/IModelWriteRpcImpl.ts
Comment (pmconne) [reply core/backend/src/rpc-impl/IModelWriteRpcImpl.ts]: Yes, dunno if worth it though.
I need to take this opportunity to say the Github PR workflow is terrible.

Comment (kabentley) [reply core/backend/src/rpc-impl/IModelWriteRpcImpl.ts]: yes, I miss ADO dearly.

File: test-apps/display-test-app/src/frontend/Viewer.ts
Comment (kabentley) [reply test-apps/display-test-app/src/frontend/Viewer.ts:409]: I made it supply the filename as the key (so it can show it on the titlebar.) So, I put openStandaloneIModel back the way it was.

File: core/backend/src/rpc-impl/RpcBriefcaseUtility.ts
Comment (wgoehrig) [inline core/backend/src/rpc-impl/RpcBriefcaseUtility.ts:109]: Should we skip this entirely if `process.env.BLOCKCACHE_DIR` is not set?

File: core/backend/src/CheckpointManager.ts
Comment (kabentley) [reply core/backend/src/CheckpointManager.ts:51]: I'm working on that. That's to provide a way to open briefcases/checkpoints that were downloaded by a previous version. Stay tuned.

File: core/backend/src/IModelDb.ts
Comment (kabentley) [reply core/backend/src/IModelDb.ts:2281]: It will. `super.findByKey` calls `this.tryFindByKey` which does the instanceOf test.

Comment (kabentley) [reply core/backend/src/IModelDb.ts:2291]: added test [here](https://github.com/imodeljs/imodeljs/blob/5c315ef2fb575bf4241acaad89014d7c63916079/core/backend/src/test/standalone/IModel.test.ts#L2035) to show that it does.

File: core/backend/src/CheckpointManager.ts
Comment (kabentley) [reply core/backend/src/CheckpointManager.ts:61]: This is now implemented.

File: core/backend/src/rpc-impl/RpcBriefcaseUtility.ts
Comment (kabentley) [reply core/backend/src/rpc-impl/RpcBriefcaseUtility.ts:129]: I don't think that's worth it.

File: docs/changehistory/NextVersion.md
Comment (pmconne) [inline docs/changehistory/NextVersion.md:185]: Links to internal or alpha APIs are invalid - will break docs build when we go to release 2.11.0.

File: editor/common/src/EditorProps.ts
Comment (pmconne) [inline editor/common/src/EditorProps.ts:10]: No docs for any of these types?

File: editor/common/src/EditorProps.ts
Comment (pmconne) [inline editor/common/src/EditorProps.ts:23]: Can a CommandResult have both an error and a result?

File: editor/backend/src/EditCommand.ts
Comment (pmconne) [inline editor/backend/src/EditCommand.ts:29]: Can you document what these do and when they are called? Difference between onCleanup and onFinish?

**Review by pmconne — CHANGES_REQUESTED** [review]

File: docs/changehistory/NextVersion.md
Comment (kabentley) [reply docs/changehistory/NextVersion.md]: done

File: core/backend/src/IModelDb.ts
Comment (pmconne) [inline core/backend/src/IModelDb.ts:179]: nitpick but does it actually throw? Looks like it just returns.

File: editor/common/src/EditorProps.ts
Comment (pmconne) [inline editor/common/src/EditorProps.ts:36]: Is there some enforcement in the RPC layer that `args` is a POD javascript value?

Comment (swbsi) [reply editor/common/src/EditorProps.ts:36]: Not at runtime. There was a custom tslint rule that enforced it at compile time. That may be disabled now with eslint though?

Comment (kabentley) [reply editor/common/src/EditorProps.ts:36]: This api doesn't use our RPC layer. Instead it uses the Electron IPC directly. Their documentation says:

Arguments will be serialized with the [Structured Clone Algorithm,](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm)

Comment (pmconne) [reply editor/common/src/EditorProps.ts:36]: Probably should document that and note that objects must be handled with care (e.g., no `instanceof`, no functions, etc).

File: editor/common/src/EditorProps.ts
Comment (kabentley) [reply editor/common/src/EditorProps.ts:10]: I'm not going to bother to document any of this until Brien and I spend more time fleshing it out. The only reason it's in this PR is to get the package.json files included so I don't have to edit them every time I merge. I started working on this and realized that I couldn't make it work until I straightened out BriefcaseDb.

Do you think it'd be clearer that none of this is stable if I marked it @alpha rather than @internal?

File: core/backend/src/IModelDb.ts
Comment (kabentley) [reply core/backend/src/IModelDb.ts]: I changed it to use Guid's for keys and throw an exception if you provide a key and it's already in use. Allowing multiple connections for the same IModelDb is a disaster waiting to happen.

File: core/backend/src/IModelDb.ts
Comment (kabentley) [reply core/backend/src/IModelDb.ts]: fixed.

File: editor/common/src/EditorProps.ts
Comment (pmconne) [reply editor/common/src/EditorProps.ts:10]: Fair enough. Yes, `internal` for APIs we don't intend others to use *ever*; `alpha` for APIs we don't intend others to use *yet*.

File: editor/common/src/EditorProps.ts
Comment (pmconne) [reply editor/common/src/EditorProps.ts:23]: I'd like to revisit this after you get the packages merged in. Seems like a CommandResult should be exactly one of either an error or a result.

File: core/backend/src/IModelDb.ts
Comment (pmconne) [reply core/backend/src/IModelDb.ts]: Does that break Shaun's test in SnapshotConnection.test.ts that opens two connections to the same iModel? Will it get rid of the hack I had to add for dealing with that situation when pushing events from backend to frontend?

File: full-stack-tests/core/src/frontend/standalone/SnapshotConnection.test.ts
Comment (kabentley) [inline full-stack-tests/core/src/frontend/standalone/SnapshotConnection.test.ts]: @pmconne see here

File: core/backend/src/IModelDb.ts
Comment (kabentley) [reply core/backend/src/IModelDb.ts]: Yes, see below.

File: editor/common/src/EditorProps.ts
Comment (kabentley) [reply editor/common/src/EditorProps.ts:23]: I dislike this too, I merely copied what we do on our native code in `ErrorStatusOrResult` in `NativeLibrary.ts` I doubt it will survive

File: full-stack-tests/core/src/frontend/standalone/SnapshotConnection.test.ts
Comment (pmconne) [reply full-stack-tests/core/src/frontend/standalone/SnapshotConnection.test.ts]: I see a deleted TODO above. When I navigate to that file in your diff I don't see any such line - only removal of explicit type names. Does this test currently pass or fail on your branch?

File: common/api/imodeljs-common.api.md
Comment (scsewall) [inline common/api/imodeljs-common.api.md:3738]: Why change from GuidString --> string?

File: full-stack-tests/core/src/frontend/standalone/SnapshotConnection.test.ts
Comment (kabentley) [reply full-stack-tests/core/src/frontend/standalone/SnapshotConnection.test.ts]: yes, I deleted the TODO so you'd know that's not true any more. There is now always a unique key for every IModelDb. Where is your hack?

File: common/api/imodeljs-common.api.md
Comment (kabentley) [reply common/api/imodeljs-common.api.md:3738]: No clue. It's GuidString in the real code, just extract-api reduced it to string for some reason.

File: full-stack-tests/core/src/frontend/standalone/SnapshotConnection.test.ts
Comment (pmconne) [reply full-stack-tests/core/src/frontend/standalone/SnapshotConnection.test.ts]: [Here](https://github.com/imodeljs/imodeljs/blob/816a34a8faa9e44c4830ee5aea443b3ffa6ed61d/core/backend/src/IModelDb.ts#L164)

Comment (kabentley) [reply full-stack-tests/core/src/frontend/standalone/SnapshotConnection.test.ts]: hack removed

**Review by pmconne — APPROVED** [review]
We should add the editor directory to CODEOWNERS soon.

**Review by bbastings — APPROVED** [review]
Edit tool/command looks good.

**Review by khanaffan — APPROVED** [review]

**Review by calebmshafer — APPROVED** [review]
Approving the addition of the new packages.

I have some questions about the EditCommand/EditTools w.r.t being able to register them during runtime but will follow-up after this PR.  Need to think about it a bit more.

**Review by scsewall — APPROVED** [review]

**Review by bsteinbk — APPROVED** [review]

**Review by swwilson-bsi — APPROVED** [review]

**Review by grigasp — APPROVED** [review]

**Review by ramanujam-raman — APPROVED** [review]

**Pull request merged**
